src/Data/UTxO/Transaction/Cardano/Shelley.hs

```diff
@@ -0,0 +1,386 @@
+-- inputs and outputs and the sorts of witnesses. Please, notice that this
+-- method overetimates the minimum acceptable fees.
+--
+-- __examples__:
+--
+-- >>> estimateFee 1 2 1 0 Cardano.Mainnet
+-- Lovelace 168097
+--
+-- >>> estimateFee 1 2 1 0 (Cardano.Testnet (Cardano.NetworkMagic 12345))
+-- Lovelace 168097
+--
+-- estimateFee 1 2 0 1 Cardano.Mainnet
+-- Lovelace 169725
+-- @since 2.0.0
+estimateFee
+    :: Int
+        -- ^ Number of inputs in tx
+    -> Int
+        -- ^ Number of outputs in tx
+    -> Int
+        -- ^ Number of Shelley witnesses in tx
+    -> Int
+        -- ^ Number of Byron witnesses in tx
+    -> NetworkId
+        -- ^ Network id
+    -> Cardano.Lovelace
+estimateFee nInps nOuts nShelleyWits nByronWits net =
+    Cardano.estimateTransactionFee net minFeeB minFeeA dummyTx nInps nOuts nShelleyWits nByronWits
+  where
+      ttl = SlotNo 10000
+      initialSigData = Cardano.makeShelleyTransaction
+          TxExtraContent
+          { txMetadata = Nothing
+          , txWithdrawals = []
+          , txCertificates = []
+          , txUpdateProposal = Nothing
+          }
+          ttl
+          (Cardano.Lovelace $ naturalToInteger minFeeB)
+          []
+          []
+      dummyTx = Cardano.makeSignedTransaction [] initialSigData
+
+--
+-- MkPayment instance
+--
+
+-- Type-level constructor capturing types for 'Shelley'.
+data Shelley
+
+-- Minimum possible value for each output
+minUTxOvalue :: Cardano.Lovelace
+minUTxOvalue = Cardano.Lovelace 1000000
+
+-- Values to calculate fee : b + a*txsize
+minFeeA, minFeeB :: Natural
+minFeeA = 44
+minFeeB = 155381
+
+type ByronSigningKey = (ByteString, SigningKey)
+
+instance MkPayment Shelley where
+    type Init Shelley = (NetworkId, SlotNo, Cardano.Lovelace)
+
+    type Input   Shelley = TxIn
+    type Output  Shelley = TxOut Cardano.Shelley
+    type SignKey Shelley = Either ByronSigningKey Cardano.ShelleyWitnessSigningKey
+
+    type CoinSel Shelley =
+        (NetworkId, SlotNo, Cardano.Lovelace, [TxIn], [TxOut Cardano.Shelley])
+
+    type Tx Shelley = Either
+        ErrMkPayment
+        ( NetworkId
+        , [TxIn]
+        , [TxOut Cardano.Shelley]
+        , Cardano.TxBody Cardano.Shelley
+        , [Cardano.Witness Cardano.Shelley]
+        )
+
+    empty :: Init Shelley -> CoinSel Shelley
+    empty (net, ttl, fee) = (net, ttl, fee, mempty, mempty)
+
+    addInput :: TxIn -> CoinSel Shelley -> CoinSel Shelley
+    addInput inp (pm, ttl, fee, inps, outs) = (pm, ttl, fee, inp : inps, outs)
+
+    addOutput :: TxOut Cardano.Shelley -> CoinSel Shelley -> CoinSel Shelley
+    addOutput out (pm, ttl, fee, inps, outs) = (pm, ttl, fee, inps, out : outs)
+
+    lock :: CoinSel Shelley -> Tx Shelley
+    lock (_net, _ttl, _fee, [], _outs) = Left MissingInput
+    lock (_net, _ttl, _fee, _inps, []) = Left MissingOutput
+    lock (net, ttl, fee, inps, outs)
+        | any (\(TxOut _ coin) -> coin < minUTxOvalue) outs = Left TooLowOutput
```

Comment:
This value depends on protocol parameters, we can't possibly know that in this library. The caller has to be aware. This library should generate transaction that are structurally well-formed, but they may still be interpreted as invalid if the network deems them so. Very much we can't verify that the provided UTxO exists or has the right amount, we can't also verify the minimum UTxO amount.

Reply:
fixed in https://github.com/input-output-hk/cardano-transactions/pull/17/commits/51ce57e3a1225b95303143a429a4a4d98b9a501b